src/backend/metalll/src/command.rs

```
@@ -0,0 +1,613 @@
+                }
+            }
+
+            let blit_encoder = self.command_buffer.new_blit_command_encoder(); // Returns retained
+            self.encoder_state = EncoderState::Blit(blit_encoder);
+            blit_encoder
+        }
+    }
+}
+
+impl core::CommandBuffer for CommandBuffer {
+    type SubmitInfo = SubmitInfo;
+
+    unsafe fn end(&mut self) -> SubmitInfo {
+        match self.encoder_state {
+            EncoderState::None => {},
+            EncoderState::Blit(blit_encoder) => blit_encoder.end_encoding(),
+            EncoderState::Render(render_encoder) => render_encoder.end_encoding(),
+        }
+
+        self.command_buffer.retain();
+        SubmitInfo {
+            command_buffer: self.command_buffer,
+        }
+    }
+}
+
+impl core::PrimaryCommandBuffer<Resources> for CommandBuffer {
+    fn pipeline_barrier<'a>(&mut self, memory_barriers: &[memory::MemoryBarrier], buffer_barriers: &[memory::BufferBarrier<'a, Resources>], image_barriers: &[memory::ImageBarrier<'a, Resources>]) {
+        //unimplemented!() FIXME
+    }
+
+    fn execute_commands(&mut self) {
+        unimplemented!()
+    }
+}
+
+impl core::GraphicsCommandBuffer<Resources> for CommandBuffer {
+    fn clear_depth_stencil(&mut self, depth_view: &native::DepthStencilView, depth_value: Option<target::Depth>, stencil_value: Option<target::Stencil>) {
+        unimplemented!()
+    }
+
+    fn resolve_image(&mut self) {
+        unimplemented!()
+    }
+
+    fn bind_index_buffer(&mut self, view: IndexBufferView<Resources>) {
+        unimplemented!()
+    }
+
+    fn bind_vertex_buffers(&mut self, buffer: pso::VertexBufferSet<Resources>) {
+        self.vertex_buffers.clear();
+        self.vertex_buffers.extend(buffer.0.iter().map(|&(buffer, offset)| (buffer.0, offset)));
+    }
+
+    fn set_viewports(&mut self, rects: &[target::Rect]) {
+        if rects.len() != 1 {
+            panic!("Metal supports only one viewport");
```

We could propagate the errors in the future and emit then on `finish()`